### ISSUE TITLE
Sending a heartbeat took longer than 2 seconds

### ISSUE THREAD

**Issue by JensForstmann** [arangodb/arangodb#12881]
## My Environment

* __ArangoDB Version__: 3.7.2-1 (Enterprise evaluation)       
* __Storage Engine__: RocksDB            
* __Deployment Mode__: Cluster       
* __Deployment Strategy__: Kubernetes  
* __Configuration__:               
* __Infrastructure__: Azure              
* __Operating System__: Ubuntu 16.04.6 LTS       
* __Total RAM in your machine__:  7Gb      
* __Disks in use__: SSD       
* __Used Package__: Docker              

Running a 3 node Kubernetes Cluster (3 agents, 3 coordinators, 3 primaries).

The nodes or the cluster in genreal are not under any load also when accessing a terminal in the containers they all react quickly without any interruption or delay.

## Component, Query & Data

__Replication Factor & Number of Shards (Cluster only)__:
Running a OneShard environment

## Steps to reproduce

1. Watch the logs on the coordinators and primaries
2. See the following messages in all of them: "2020-10-20T15:25:41.670788465Z 2020-10-20T15:25:41Z [1] WARNING [77655] {heartbeat} ATTENTION: Sending a heartbeat took longer than 2 seconds, this might be causing trouble with health checks. Please contact ArangoDB Support."

[example log primary 1.txt](https://github.com/arangodb/arangodb/files/5409866/example.log.primary.1.txt)

__Problem__:

That's an obvious warning that this might cause some trouble ;-)

Not sure if that has something to do with that but I noticed the followting two issues but unfortunately couldn't reproduce them reliable:

1. Unique constraint violation on the _key attribute when inserting new documents within a transaction (which is strange because we do not set them explicitly). This happened several times now and it seems to stuck in that case (always aborting the transaction with the same error) for a couple of minutes and then resolves without any action taken.

2. Transactions were aborted when trying to delete documents with the message that one of the primaries went down but the logs of all primaries didn't had any error messages in them and also the pods were not restarted.

__Expected result__:

Not having the warnings about the heartbeats (and hopefully not having transactions which abort without any clear reason from my point of view).

**Comment by curtdept** [comment]
Azure / Kubernetes has very well known DNS weaknesses (CONNTRACK races, long UDP timeouts, etc). Use this workaround, it deploys nodelocaldns on Azure.

https://raw.githubusercontent.com/curtdept/aks_nodelocaldns/main/nodelocaldns.yaml

Some more info: https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/

**Comment by JensForstmann** [comment]
Ok, I'll give it a try and report how it went. But could take a couple of days (prob. weeks).

**Comment by curtdept** [comment]
@JensForstmann it's worth it for all your workloads. Kubernetes slow DNS has been a multi-year problem.

**Comment by dothebart** [comment]
@curtdept thanks for this pointer about DNS.
We will work on have better error reporting and more statistics here, so similar issues can be obvious to users in the future.

**Comment by JensForstmann** [comment]
Today we deployed as @curtdept suggested nodelocaldns in our AKS cluster.

Unfortunately that didn't change anything: We still receive heartbeat timeout log messages.

Any other ideas?

**Comment by dothebart** [comment]
whats the ping between the systems? whats the load?

**Comment by curtdept** [comment]
What's AKS version? Some of the version cause the local dns containers to get killed and restart.

You may have to remove the labels 
kubernetes.io/cluster-service: "true" addonmanager.kubernetes.io/mode: Reconcile

In that plan to prevent it.

Also are you kubenet or azure cni?

**Comment by JensForstmann** [comment]
AKS version 1.16.9
Azure CNI

I removed all labels with kubernetes.io/cluster-service: "true" and addonmanager.kubernetes.io/mode: Reconcile. Let's see what happens. (It seems that there were several restarts, yes.)

Pings are between 0.031 ms and 6.644 ms (avg 1.390 ms)

**Comment by curtdept** [comment]
@JensForstmann how did it turn out?

**Comment by JensForstmann** [comment]
Hi and sorry for the delayed response.

I redeployed your linked yaml file (nodelocadns.yaml) since the pods were gone (but not sure if we removed them or AKS did that).

At least now there are no labels like `kubernetes.io/cluster-service: "true"` or `addonmanager.kubernetes.io/mode: Reconcile`

And I see four node-local-dns pods runing, one on every node.


The logs from servers (prmr) still contain the heatbeat log warnings though. Together with multiple info messages that both prmr and crdn are gone or rebooted.

I saved the logs from the complete cluster (operators, agents, coordinators, primaries) and attached it. Maybe that helps more than my initial logfile. The log contains the output of 10 minutes. The ArangoDB cluster was completely idle and empty, only one webui was opened.

[log_2021-01-13_14-40.txt](https://github.com/arangodb/arangodb/files/5809333/log_2021-01-13_14-40.txt)

**Comment by dothebart** [comment]
from that log: 
```
dial tcp: lookup forklift.basket.svc on 10.0.0.10:53: no such host
```
port 53 is DNS. So it seems your DNS doesn't work properly?

**Comment by JensForstmann** [comment]
### tl;dr
Lot's of stuff and terminal output below collected during testing blow.

**tl;dr:**

`forklift.basket.svc` cannot be resolved, `forklift.basket.svc.cluster.local` CAN be resolved

```
/ # nslookup forklift.basket.svc.cluster.local
Server:         10.0.0.10
Address:        10.0.0.10:53


Name:   forklift.basket.svc.cluster.local
Address: 10.0.115.174

/ # nslookup forklift.basket.svc.cluster.local
Server:         10.0.0.10
Address:        10.0.0.10:53


Name:   forklift.basket.svc.cluster.local
Address: 10.0.115.174

/ #
```

### long version

```
$ kubectl --namespace basket get pods -o wide
NAME                                          READY   STATUS    RESTARTS   AGE    IP             NODE                                NOMINATED NODE   READINESS GATES
arango-deployment-operator-6f6b646f85-8ntrm   1/1     Running   0          23h    10.240.0.22    aks-agentpool-83463867-vmss000000   <none>           <none>
arango-deployment-operator-6f6b646f85-xln58   1/1     Running   0          23h    10.240.0.121   aks-agentpool-83463867-vmss000004   <none>           <none>
forklift-agnt-9ccigee7-cd75a4                 1/1     Running   0          22h    10.240.0.96    aks-agentpool-83463867-vmss000002   <none>           <none>
forklift-agnt-cfwc9agv-cd75a4                 1/1     Running   0          109m   10.240.0.76    aks-agentpool-83463867-vmss000002   <none>           <none>
forklift-agnt-vfsfkv5s-cd75a4                 1/1     Running   0          22h    10.240.0.56    aks-agentpool-83463867-vmss000001   <none>           <none>
forklift-crdn-gactzczd-cd75a4                 1/1     Running   0          22h    10.240.0.110   aks-agentpool-83463867-vmss000004   <none>           <none>
forklift-crdn-ib6z15w0-cd75a4                 1/1     Running   0          22h    10.240.0.5     aks-agentpool-83463867-vmss000000   <none>           <none>
forklift-crdn-zqxnqlbr-cd75a4                 1/1     Running   0          22h    10.240.0.40    aks-agentpool-83463867-vmss000001   <none>           <none>
forklift-prmr-ma6dv4yy-cd75a4                 1/1     Running   0          22h    10.240.0.63    aks-agentpool-83463867-vmss000001   <none>           <none>
forklift-prmr-nebapsza-cd75a4                 1/1     Running   0          22h    10.240.0.21    aks-agentpool-83463867-vmss000000   <none>           <none>
forklift-prmr-qm4qrdex-cd75a4                 1/1     Running   0          22h    10.240.0.77    aks-agentpool-83463867-vmss000002   <none>           <none>

```

The messages show probelsm from `arango-deployment-operator-6f6b646f85-xln58` so i logged into a pod on the same node:

```
$ kubectl --namespace basket exec -it forklift-crdn-gactzczd-cd75a4 -- sh

/ # nslookup forklift.basket.svc

Server:         10.0.0.10
Address:        10.0.0.10:53

** server can't find forklift.basket.svc: NXDOMAIN

** server can't find forklift.basket.svc: NXDOMAIN

/ # nslookup forklift.basket.svc
Server:         10.0.0.10
Address:        10.0.0.10:53

** server can't find forklift.basket.svc: NXDOMAIN

** server can't find forklift.basket.svc: NXDOMAIN

/ #
```

So there seems to be a dns problem, yes. Tested this on other pods (on other nodes) and there it's the same.

I've tested the same in another namespace and the situation was the same. Afterdeleting the nodelocaldns resources the lookups sometimes were good, see:


```
/ # nslookup forklift.basket.svc.cluster.local
Server:         10.0.0.10
Address:        10.0.0.10:53


Name:   forklift.basket.svc.cluster.local
Address: 10.0.115.174

/ # nslookup forklift.basket.svc.cluster.local
Server:         10.0.0.10
Address:        10.0.0.10:53

** server can't find forklift.basket.svc.cluster.local: NXDOMAIN

** server can't find forklift.basket.svc.cluster.local: NXDOMAIN

/ #
```

Executed directly afterwards, in about 3 out of 10 I got a positive (but also sometimes slow) response.

But I was never able to get a response for example "forklift-agent-9ccigee7.forklift-int.basket.svc" or "forklift-dbserver-qm4qrdex.forklift-int.basket.svc" which failed requests you see in the log. But "forklift-dbserver-qm4qrdex.forklift-int.basket.svc.cluster.local" is working (even with nodelocaldns).

It seems that the suffix ".cluster.local" is important to resolve the names correctly.

```
/ # cat /etc/resolv.conf
nameserver 10.0.0.10
search basket.svc.cluster.local svc.cluster.local cluster.local 0ycwsc1wo5jelof4sssmadduog.ax.internal.cloudapp.net
options ndots:5
```

**Comment by JensForstmann** [comment]
It seems that only ArangoDB pods fail to resolve names. Other pods on the same node don't fail resolving the names.

From https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/ I deployed the dnsutils pod.

```
$ kubectl --namespace basket get pods -o wide
NAME                                          READY   STATUS    RESTARTS   AGE   IP             NODE                                NOMINATED NODE   READINESS GATES
arango-deployment-operator-6f6b646f85-8ntrm   1/1     Running   0          45h   10.240.0.22    aks-agentpool-83463867-vmss000000   <none>           <none>
arango-deployment-operator-6f6b646f85-xln58   1/1     Running   0          45h   10.240.0.121   aks-agentpool-83463867-vmss000004   <none>           <none>
dnsutils                                      1/1     Running   0          10s   10.240.0.81    aks-agentpool-83463867-vmss000002   <none>           <none>
forklift-agnt-9ccigee7-cd75a4                 1/1     Running   0          44h   10.240.0.96    aks-agentpool-83463867-vmss000002   <none>           <none>
forklift-agnt-cfwc9agv-cd75a4                 1/1     Running   0          23h   10.240.0.76    aks-agentpool-83463867-vmss000002   <none>           <none>
forklift-agnt-vfsfkv5s-cd75a4                 1/1     Running   0          44h   10.240.0.56    aks-agentpool-83463867-vmss000001   <none>           <none>
forklift-crdn-gactzczd-cd75a4                 1/1     Running   0          44h   10.240.0.110   aks-agentpool-83463867-vmss000004   <none>           <none>
forklift-crdn-ib6z15w0-cd75a4                 1/1     Running   0          43h   10.240.0.5     aks-agentpool-83463867-vmss000000   <none>           <none>
forklift-crdn-zqxnqlbr-cd75a4                 1/1     Running   0          44h   10.240.0.40    aks-agentpool-83463867-vmss000001   <none>           <none>
forklift-prmr-ma6dv4yy-cd75a4                 1/1     Running   0          44h   10.240.0.63    aks-agentpool-83463867-vmss000001   <none>           <none>
forklift-prmr-nebapsza-cd75a4                 1/1     Running   0          44h   10.240.0.21    aks-agentpool-83463867-vmss000000   <none>           <none>
forklift-prmr-qm4qrdex-cd75a4                 1/1     Running   0          44h   10.240.0.77    aks-agentpool-83463867-vmss000002   <none>           <none>
```

I picked `forklift-agnt-9ccigee7-cd75a4` as a pod to test with to have the same node as dnsutils:

```
$ kubectl --namespace basket exec -it forklift-agnt-9ccigee7-cd75a4 -- cat /etc/resolv.conf
nameserver 10.0.0.10
search basket.svc.cluster.local svc.cluster.local cluster.local 0ycwsc1wo5jelof4sssmadduog.ax.internal.cloudapp.net
options ndots:5
```

```
$ kubectl --namespace basket exec -it dnsutils -- cat /etc/resolv.conf
nameserver 10.0.0.10
search basket.svc.cluster.local svc.cluster.local cluster.local 0ycwsc1wo5jelof4sssmadduog.ax.internal.cloudapp.net
options ndots:5
```

```
$ kubectl --namespace basket exec -it forklift-agnt-9ccigee7-cd75a4 -- nslookup -debug forklift.basket.svc
Server:         10.0.0.10
Address:        10.0.0.10:53

Query #0 completed in 162ms:
** server can't find forklift.basket.svc: NXDOMAIN

Query #1 completed in 162ms:
** server can't find forklift.basket.svc: NXDOMAIN

```

```
$ kubectl --namespace basket exec -it dnsutils -- nslookup -debug forklift.basket.svc
Server:         10.0.0.10
Address:        10.0.0.10#53

------------
    QUESTIONS:
        forklift.basket.svc.basket.svc.cluster.local, type = A, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ->  cluster.local
        origin = ns.dns.cluster.local
        mail addr = hostmaster.cluster.local
        serial = 1610633105
        refresh = 7200
        retry = 1800
        expire = 86400
        minimum = 5
        ttl = 5
    ADDITIONAL RECORDS:
------------
** server can't find forklift.basket.svc.basket.svc.cluster.local: NXDOMAIN
Server:         10.0.0.10
Address:        10.0.0.10#53

------------
    QUESTIONS:
        forklift.basket.svc.svc.cluster.local, type = A, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ->  cluster.local
        origin = ns.dns.cluster.local
        mail addr = hostmaster.cluster.local
        serial = 1610633105
        refresh = 7200
        retry = 1800
        expire = 86400
        minimum = 5
        ttl = 5
    ADDITIONAL RECORDS:
------------
** server can't find forklift.basket.svc.svc.cluster.local: NXDOMAIN
Server:         10.0.0.10
Address:        10.0.0.10#53

------------
    QUESTIONS:
        forklift.basket.svc.cluster.local, type = A, class = IN
    ANSWERS:
    ->  forklift.basket.svc.cluster.local
        internet address = 10.0.115.174
        ttl = 5
    AUTHORITY RECORDS:
    ADDITIONAL RECORDS:
------------
Name:   forklift.basket.svc.cluster.local
Address: 10.0.115.174
------------
    QUESTIONS:
        forklift.basket.svc.cluster.local, type = AAAA, class = IN
    ANSWERS:
    AUTHORITY RECORDS:
    ->  cluster.local
        origin = ns.dns.cluster.local
        mail addr = hostmaster.cluster.local
        serial = 1610633105
        refresh = 7200
        retry = 1800
        expire = 86400
        minimum = 5
        ttl = 5
    ADDITIONAL RECORDS:
------------
```


And that is the file for the ArangoDeployment which I use here for testing:

```yaml
apiVersion: "database.arangodb.com/v1alpha"
kind: "ArangoDeployment"
metadata:
  name: "forklift"
  namespace: basket
spec:
  mode: Cluster
  image: arangodb/enterprise:3.7.6
  license:
    secretName: arango-license-key
  tls:
    altNames: ["forklift.basket.k8s.vdb.st"]

```

Any ideas what else I could do to track down the issue?

**Comment by JensForstmann** [comment]
@curtdept You mentioned running Arango also in AKS. How did you manage to get the ArangoDB docker image to resolve the non FQDN requests? The alpine base image is ignoring the `ndots:5` option and thus only makes a single request when there is at least one dot in the name (assuming it's a FQDN, even if it isn't).

**Comment by curtdept** [comment]
@JensForstmann ya, no issue in mine

I did find this though, maybe it applies?
https://github.com/Azure/AKS/issues/2052

other then that, the only other thing I do explicitly is disableIPv6: true in my cluster deployment.

**Comment by JensForstmann** [comment]
Azure/AKS#2052 doesn't apply since I don't have a custom DNS server.

@curtdept What version of ArangoDB are you using? And does a `nslookup test.test` from within the ArangoDB pod really request all search domains from `/etc/resolv.conf`?

I'll will give it a try to disable ipv6 though

**Comment by JensForstmann** [comment]
`disableIPv6: true` didn't change the situation.

But I upgraded the cluster to 1.18.14 and the node pool from version 1.16.x to 1.17.16 and it seems that this resolves this mystery. No more timeouts or reolving issues for about an hour. 👍

I'll close this issue.

@curtdept Just curious: On what version are you for both cluster and node pool in AKS? I see that I could upgrade to even higher versions.

**Comment by curtdept** [comment]
@JensForstmann 1.19.6 (latest)
